Win32.xs

```diff
@@ -1682,6 +1685,300 @@ XS(w32_IsDeveloperModeEnabled)
+    if (bResults) {
+        hOut = CreateFileW(file,
+                           GENERIC_WRITE,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE,
+                           NULL,
+                           CREATE_ALWAYS,
+                           FILE_ATTRIBUTE_NORMAL,
+                           NULL);
+
+        if (hOut == INVALID_HANDLE_VALUE)
+            bFileError = TRUE;
+    }
+
+    if (!bFileError && bResults) {
+        DWORD dwDownloaded = 0;
+        DWORD dwBytesWritten = 0;
+        DWORD dwSize = 65536;
+        char *pszOutBuffer;
+
+        New(0, pszOutBuffer, dwSize, char);
+
+        /* Keep checking for data until there is nothing left. */
+        while (1) {
+            if (!WinHttpReadData(hRequest,
+                                 (LPVOID)pszOutBuffer,
+                                 dwSize,
+                                 &dwDownloaded)) {
+                bAborted = TRUE;
+                break;
+            }
+            if (!dwDownloaded)
+                break;
+
+            /* Write what we just read to the output file */
+            if (!WriteFile(hOut,
+                           pszOutBuffer,
+                           dwDownloaded,
+                           &dwBytesWritten,
+                           NULL)) {
+                bAborted = TRUE;
+                bFileError = TRUE;
+                break;
+            }
+
+        }
+
+        Safefree(pszOutBuffer);
+    }
+    else {
+        bAborted = TRUE;
+    }
+
+    /* Clean-up may lose this. */
+    if (bAborted)
+        error = GetLastError();
+
+    /* Close any open handles. */
+    if (hOut) CloseHandle(hOut);
+    if (hRequest) WinHttpCloseHandle(hRequest);
+    if (hConnect) WinHttpCloseHandle(hConnect);
+    if (hSession) WinHttpCloseHandle(hSession);
+
+    Safefree(url);
+    Safefree(file);
+    Safefree(hostName);
+    Safefree(urlPath);
+
+    /* Retrieve system and WinHttp error messages, but not if we already
+     * got a failed HTTP status text above.
+     */
+    if (bAborted && !bHttpError) {
+        DWORD msgFlags = bFileError
+                         ? FORMAT_MESSAGE_FROM_SYSTEM
+                         : FORMAT_MESSAGE_FROM_HMODULE;
+        msgFlags |= FORMAT_MESSAGE_IGNORE_INSERTS;
+
+        ZeroMemory(&msgbuf, ONE_K_BUFSIZE * 2);
+        if (!FormatMessageW(msgFlags,
+                            GetModuleHandleW(L"winhttp.dll"),
+                            error,
+                            0,
+                            msgbuf,
+                            ONE_K_BUFSIZE - 1, /* TCHARs, not bytes */
+                            NULL)) {
+            wcsncpy(msgbuf, L"unable to format error message", ONE_K_BUFSIZE - 1);
+        }
+        SetLastError(error);
+    }
+
+    if (GIMME_V == G_SCALAR) {
+        EXTEND(SP, 1);
+        ST(0) = !bAborted ? &PL_sv_yes : &PL_sv_no;
+        XSRETURN(1);
+    }
+    else if (GIMME_V == G_ARRAY) {
+        EXTEND(SP, 2);
```

Comment:
I think the `EXTEND` is not needed because you have 2 mandatory arguments, so `ST(0)` and `ST(1)` are always available without moving the stack around. But I guess having the `EXTEND` doesn't hurt, even if it is just a no-op.

Reply:
> I think the `EXTEND` is not needed because you have 2 mandatory arguments, so `ST(0)` and `ST(1)` are always available without moving the stack around. But I guess having the `EXTEND` doesn't hurt, even if it is just a no-op.

perlxstut says, "it's not a good idea to write extensions that modify their input parameters," which I thought meant reusing anything in the ST array without extending, but I don't claim to have a very good understanding of XS stack handling.

I think I have added all the requested features and tests, but there are a few things that might merit further review:

- The "use POSIX" statement in the test script throws a compile-time warning but does not prevent the tests from succeeding.  Surely that's a bug in POSIX?  What I'm trying to do here is set the locale so the error messages can be compared to known values for testing purposes, but I have no idea whether locales even work on Windows, and no non-English locales to test with.

- I cannot add a test for ignoring invalid certs because I don't know of a persistent internet location with a bad certificate that is going to stay that way.  A manual test with a local server having a self-signed certificate seems to do the right thing.

- I cannot test the auto proxy handling code since I don't have any Windows machines on networks that have proxies or the access or know-how to add a proxy to a Windows network.

Reply:
> perlxstut says, "it's not a good idea to write extensions that modify their input parameters," which I thought meant reusing anything in the ST array without extending

You are not re-using the input parameters. You are putting new/different SVs in the same position on the stack. If you are called with 2 parameters, than you know you can return 2 values on the stack without having to extend it. If you want to return 3, then you need to `EXTEND(SP, 3)` to make sure `ST(2)` is not outside the bounds of the allocated stack.

Just providing the information here as an explanation; I think leaving in the `EXTEND` is fine, and probably more maintenance friendly...

Reply:
> I cannot add a test for ignoring invalid certs because I don't know of a persistent internet location with a bad certificate that is going to stay that way.

You can probably use one of the [chromium/badssl.com: Memorable site for testing clients against bad SSL configs.](https://github.com/chromium/badssl.com) domains for this.

Reply:
> I cannot test the auto proxy handling code since I don't have any Windows machines on networks that have proxies or the access or know-how to add a proxy to a Windows network.

I will probably have to do some work with Windows proxy setups in the future, so will try to test it manually when I have something. I'm fine with leaving it untested for now; we can fix it later, if it turns out to be broken.

Reply:
> The "use POSIX" statement in the test script throws a compile-time warning but does not prevent the tests from succeeding. Surely that's a bug in POSIX? 

No, it is not a bug; it is a conflict with the Win32 module. For historic reasons Win32 exports a `NULL` function without a prototype:

```perl
sub NULL 				{ 0 }
```

However, the `locale_h` group of exports from POSIX also exports a `NULL` function, alas with a proper `()` prototype to denote it as a constant.

Newer constants in Win32 use the proper prototype, but for the original constants we couldn't change it because it would create warnings with other Win32 modules (I think; it has been a long time, but I know I tried to use prototypes for all).

Anyways, you can get rid of the warning by not importing anything from the Win32 package:

```perl
use Win32 qw();
```

I don't think switching the locale will have any effect on the error messages, but I still have to test it (not sure if I can, but I'll try).

Reply:
> What I'm trying to do here is set the locale so the error messages can be compared to known values for testing purposes, but I have no idea whether locales even work on Windows, and no non-English locales to test with.

I've tested this now, and it doesn't work; you still get the localized error message. You need to change the `dwLanguageId` parameter to `FormatMessage` to the desired language, or call `SetThreadLocale()` to set the Windows locale for the current thread. I've confirmed that passing `0x409` (the language id for US English) will return the English error, even when `0` returns the localized error.

This is documented in the [FormatMessage documentation](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessage):

> If you pass a specific LANGID in this parameter, FormatMessage will return a message for that LANGID only. If the function cannot find a message for that LANGID, it sets Last-Error to ERROR_RESOURCE_LANG_NOT_FOUND. If you pass in zero, FormatMessage looks for a message for LANGIDs in the following order:
> 1. Language neutral
> 2. Thread LANGID, based on the thread's locale value
> 3. User default LANGID, based on the user's default locale value
> 4. System default LANGID, based on the system default locale value
> 5. US English
>
> If FormatMessage does not locate a message for any of the preceding LANGIDs, it returns any language message string that is present. If that fails, it returns ERROR_RESOURCE_LANG_NOT_FOUND.

The POSIX locale only affects the C runtime library, not the Windows API.

I would check that the current locale is English, and just skip the test if it isn't. Not sure how to check the locale with core Perl functions; I would probably do:

```perl
Win32::SetLastError(1);
if ($^E eq "Incorrect function") {
    ok(...);
} else {
    skip("Cannot verify error on non-English locale setting");
}
```